### DISCUSSION TITLE
How to upload to services like Amazon S3, Google Cloud Storage, etc. in v5?

### DISCUSSION THREAD

**Discussion by D-Marc1** [adonisjs/core#1545] [Help]
Since the File Storage (aka Drive) package is [not yet ported over to v5](https://trello.com/c/o2vckACA/16-file-storage-aka-drive), is there any way to upload files to a service like Amazon S3, Google Cloud Storage, etc? The [v5 docs state that you can stream files](https://preview.adonisjs.com/guides/http/file-uploads/#file-uploads), but doesn't explain how. Is it the same as [in v4](https://adonisjs.com/docs/4.1/file-uploads#_streaming_files)?

**Comment by Reptiluka** [comment]
When I read the doc yesterday I got the same question !
In my future app I want to use Google Cloud, it will be cool !

**Comment by mcordoba** [comment]
Im trying this right now...

I have set the autoprocess: false in bodyparser config and then trying this:

```javascript
request.multipart.onFile('image', {}, async (file, reporter) => {
    
     /*
    Here file is a PassThrough and you have file.headers and file.filename properties
    but it is a bit weird as they are inside a PassThrough. In v4.1, you had file.stream
    and you had the stream on a property.
    */
})

await request.multipart.process()
```

It would be nice an explanation of this changes and if Im doing it right @thetutlage

**Reply by thetutlage** [reply]
Going to write a blog post on it over the weekend. I am going to showcase an example with AWS S3, I hope that is enough for us?

**Reply by D-Marc1** [reply]
@thetutlage You're the man, thanks! Yes, a simple example showing streaming in v5 (especially differences from v4) with AWS S3 would be perfect.

**Reply by mcordoba** [reply]
It is working for me using variable "file" as the stream

Anyway, I miss many info on the documentation. Sometimes I spend time digging into framework code and when I discover how to use something I would like to contribute to clarify it on the documentation.

**Reply by yarapolana** [reply]
@thetutlage you got it all, looking forward for the blog post

**Reply by aayush123** [reply]
@thetutlage eagerly waiting for the post!

**Reply by D-Marc1** [reply]
In retrospect, I'm not sure this even needs a dedicated blog post. It's the same as in v4, except the Drive plugin isn't ported over yet. I just use the default AWS Node SDK with the same setup as in v4 otherwise.

**Reply by aayush123** [reply]
Meaning you use aws-sdk directly instead of using adonis package?

**Reply by D-Marc1** [reply]
Correct, the adonis `Drive` package isn't ported over to v5 yet.

**Reply by aayush123** [reply]
Yeah, I was thinking of doing the same if it doesn't get ported by the time I end up needing the feature

**Reply by yarapolana** [reply]
If you guys are interested here is my code with aws-sdk
![1](https://user-images.githubusercontent.com/19730118/94816743-c4ddcc00-03eb-11eb-8e0c-0308e52248db.png)
![2](https://user-images.githubusercontent.com/19730118/94816752-c8715300-03eb-11eb-899e-76856c64fa83.png)
![3](https://user-images.githubusercontent.com/19730118/94816754-c909e980-03eb-11eb-9375-b34919929a73.png)

**Reply by yarapolana** [reply]
And how to retrieve

![4](https://user-images.githubusercontent.com/19730118/94818065-39653a80-03ed-11eb-9121-ed997262e105.png)
![5](https://user-images.githubusercontent.com/19730118/94818073-3bc79480-03ed-11eb-9796-d0a419c0ba9d.png)

**Reply by ibalaji777** [reply]
@mcordoba getting error

code: "E_RUNTIME_EXCEPTION"
message: "E_RUNTIME_EXCEPTION: multipart stream has already been consumed"

https://github.com/adonisjs/core/discussions/2958

**Reply by 0x69h** [reply]
> @mcordoba getting error
> 
> code: "E_RUNTIME_EXCEPTION" message: "E_RUNTIME_EXCEPTION: multipart stream has already been consumed"
> 
> #2958

我的到了同样的错误，请问您如何解决了？
